alg/teca_model_segmentation.py

```diff
@@ -0,0 +1,189 @@
+    """
+    A generic TECA algorithm that provides torch based deep
+    learning feature detecting algorithms with core TECA 
+    capabilities for easy integration.
+    """
+    def __init__(self):
+        self.variable_name = None
+        self.pred_name = self.variable_name + "_PRED"
+        self.var_array = None
+        self.torch_inference_fn = None
+        self.transform_fn = None
+        self.transport_fn_args = None
+        self.model = None
+        self.model_path = None
+        self.device = self.set_torch_device()
+
+    def __str__(self):
+        ms_str = 'variable_name=%s, pred_name=%d\n\n'%( \
+            self.variable_name, self.pred_name)
+
+        ms_str += 'model:\n%s\n\n'%(str(self.model))
+
+        ms_str += 'device:\n%s\n'%(str(self.device))
+
+        return ms_str
+
+    def load_state_dict(self, state_dict_file):
+        """
+        Load only the pytorch state_dict parameters file only
+        once and broadcast it to all ranks
+        """
+        comm = self.get_communicator()
+        rank = comm.Get_rank()
+
+        sd = None
+        if rank == 0:
+            sd = torch.load(state_dict_file, map_location=lambda storage, loc: storage)
+        sd = comm.bcast(sd, root=0)
+        #sd = torch.load(state_dict_file, map_location=lambda storage, loc: storage)
+        
+        return sd
+
+    def set_variable_name(self, name):
+        """
+        set the variable name that will be inputed to the model
+        """
+        self.variable_name = str(name)
+        self.set_pred_name(self.variable_name + "_pred")
+
+    def set_pred_name(self, name):
+        """
+        set the variable name that will be the output to the model
+        """
+        self.pred_name = name
+
+    def set_torch_inference_fn(self, torch_fn):
+        """
+        set the final torch inference function. ex. torch.sigmoid()
+        """
+        self.torch_inference_fn = torch_fn
+
+    def __set_transform_fn(self, fn, *args):
+        """
+        if the data need to be transformed in a way then a function
+        could be provided to be applied on the requested data before
+        running it to the model.
+        """
+        if not hasattr(fn, '__call__'):
+            raise TypeError("ERROR: The provided data transform function "
+                "is not a function")
+
+        if not args:
+            raise ValueError("ERROR: The provided data transform function "
+                "must at least have 1 argument -- the data array object to "
+                "apply the transformation on.")
+
+        self.transform_fn = fn
+        self.transport_fn_args = args
+
+    def set_torch_device(self, device="cpu"):
+        """
+        Set device to either 'cuda' or 'cpu'
+        """
+        if device == "cuda" and not torch.cuda.is_available():
+            raise Exception("ERROR: Couldn\'t set device to cuda, cuda is "
+                "not available")
+
+        return torch.device(device)
```

Comment:
what happens if a user calls this before running the pipeline, to for instance run on a GPU? No member variables of the class are modified, does the call to torch.device(device) have a side affect of sending all subsequent torch processing to that device?

Reply:
It is a torch inner configuration that tells torch to move all its tensor arrays to the chosen device even before creating those tensors. So I think no operations will occur before we start specifying the input data and loading the model itself.

Please let me know if I am missing something.
https://pytorch.org/docs/stable/tensor_attributes.html#torch.torch.device

Reply:
The issue is that in your class there is a member variable `self.device` but this function does not update it. Someone could call this function to change the device and the class meber variable is set to the old device, that could be problematic if the class were to use `self.device` it would not modify the active device. Instead of returning the device here it would be better instead to update the member variable.

Reply:
Good point